scipy/special/_logsumexp.py

```diff
@@ -100,53 +101,113 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     """
     xp = array_namespace(a, b)
     a, b = xp_broadcast_promote(a, b, ensure_writeable=True, force_floating=True, xp=xp)
-    axis = tuple(range(a.ndim)) if axis is None else axis
+    a = xp_atleast_nd(a, ndim=1)
+    b = xp_atleast_nd(b, ndim=1) if b is not None else b
 
-    if b is not None:
-        a[b == 0] = -xp.inf
-
-    # Scale by real part for complex inputs, because this affects
-    # the magnitude of the exponential.
     if xp_size(a) == 0:
-        # because `xp.max` doesn't have `initial` argument...
-        shape = np.asarray(a.shape)  # NumPy is concise for scalar or tuple `axis`
+        shape = np.asarray(a.shape)  # NumPy is convenient for shape manipulation
+        axis = tuple(range(a.ndim)) if axis is None else axis
         shape[axis] = 1
-        a_max = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        out = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        sgn = xp.sign(out)
+    else:
+        with np.errstate(divide='ignore', invalid='ignore'):  # log of zero is OK
+            out, sgn = _logsumexp(a, b, axis=axis, return_sign=return_sign, xp=xp)
+
+    # Deal with shape details - reducing dimensions and convert 0-D to scalar for NumPy
+    out = xp.squeeze(out, axis=axis) if not keepdims else out
+    sgn = xp.squeeze(sgn, axis=axis) if (sgn is not None and not keepdims) else sgn
```

Comment:
This is annoyingly complicated, but I guess we can't remove `keepdims` now. Necessary to do this for sgn? I suspect so, otherwise you wouldn't have this line here.

Reply:
> but I guess we can't remove `keepdims` now

It is easier to do the calculations internally with `keepdims=True`. The annoying thing is supporting `keepdims=False`, the default, but it is very natural for the user to want this function to behave like any other reducing operation.

It *looked* a little simpler in the old implementation, but this is how it was working - it used `keepdims` throughout (because that's much more convenient) and squeezed at the end.

Reply:
It's possible that some of this could be removed by letting the last reducing operation in `_logsumexp` eliminate the axes depending on `keepdims`. For simplicity, I just decided to ignore that possibility for now and separate these details in `logsumexp` from the math stuff in `_logsumexp`.